File: lab3/zad2/go/onelist/onelist.go
package onelist

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
)

type OneList struct {
	head *OneListNode
	tail *OneListNode
}

func NewOneList() *OneList {
	return &OneList{}
}

func (ol *OneList) AddToHead(val string) {
	newNode := &OneListNode{Data: val}
	newNode.Next = ol.head
	ol.head = newNode
	if ol.tail == nil {
		ol.tail = newNode
	}
}

func (ol *OneList) AddToTail(val string) {
	newNode := &OneListNode{Data: val}
	if ol.tail == nil {
		ol.head = newNode
		ol.tail = newNode
	} else {
		ol.tail.Next = newNode
		ol.tail = newNode
	}
}

func (ol *OneList) AddAfterValue(targetVal, newVal string) error {
	targetNode := ol.findByVal(targetVal)
	if targetNode == nil {
		return errors.New("error")
	}

	newNode := &OneListNode{Data: newVal}
	newNode.Next = targetNode.Next
	targetNode.Next = newNode

	if targetNode == ol.tail {
		ol.tail = newNode
	}
	return nil
}

func (ol *OneList) AddBeforeValue(targetVal, newVal string) error {
	targetNode := ol.findByVal(targetVal)
	if targetNode == nil {
		return errors.New("error")
	}

	if targetNode == ol.head {
		ol.AddToHead(newVal)
		return nil
	}

	prevNode := ol.head
	for prevNode != nil && prevNode.Next != targetNode {
		prevNode = prevNode.Next
	}

	if prevNode == nil {
		return errors.New("error")
	}

	newNode := &OneListNode{Data: newVal}
	newNode.Next = targetNode
	prevNode.Next = newNode
	return nil
}

func (ol *OneList) RemoveFromHead() error {
	if ol.head == nil {
		return errors.New("error")
	}
	ol.head = ol.head.Next
	if ol.head == nil {
		ol.tail = nil
	}
	return nil
}

func (ol *OneList) RemoveFromTail() error {
	if ol.head == nil {
		return errors.New("error")
	}

	if ol.head == ol.tail {
		ol.head = nil
		ol.tail = nil
		return nil
	}

	curr := ol.head
	for curr.Next != ol.tail {
		curr = curr.Next
	}
	curr.Next = nil
	ol.tail = curr
	return nil
}

func (ol *OneList) RemoveAfterValue(targetVal string) error {
	targetNode := ol.findByVal(targetVal)
	if targetNode == nil || targetNode.Next == nil {
		return errors.New("error")
	}

	temp := targetNode.Next
	targetNode.Next = temp.Next

	if temp == ol.tail {
		ol.tail = targetNode
	}
	return nil
}

func (ol *OneList) RemoveBeforeValue(targetVal string) error {
	targetNode := ol.findByVal(targetVal)
	if targetNode == nil || targetNode == ol.head {
		return errors.New("error")
	}

	if ol.head.Next == targetNode {
		return ol.RemoveFromHead()
	}

	prevPrev := ol.head
	for prevPrev.Next != nil && prevPrev.Next.Next != targetNode {
		prevPrev = prevPrev.Next
	}

	if prevPrev.Next == nil {
		return errors.New("error")
	}

	prevPrev.Next = targetNode
	return nil
}

func (ol *OneList) RemoveByValue(targetVal string) error {
	targetNode := ol.findByVal(targetVal)
	if targetNode == nil {
		return errors.New("error")
	}

	if targetNode == ol.head {
		return ol.RemoveFromHead()
	}

	prevNode := ol.head
	for prevNode != nil && prevNode.Next != targetNode {
		prevNode = prevNode.Next
	}

	if prevNode == nil {
		return errors.New("error")
	}

	prevNode.Next = targetNode.Next
	if targetNode == ol.tail {
		ol.tail = prevNode
	}
	return nil
}

func (ol *OneList) GetElementByValue(targetVal string) (string, error) {
	targetNode := ol.findByVal(targetVal)
	if targetNode == nil {
		return "", errors.New("error")
	}
	return targetNode.Data, nil
}

func (ol *OneList) findByVal(val string) *OneListNode {
	curr := ol.head
	for curr != nil {
		if curr.Data == val {
			return curr
		}
		curr = curr.Next
	}
	return nil
}

func (ol *OneList) Print() {
	if ol.head == nil {
		fmt.Println("error")
		return
	}

	curr := ol.head
	fmt.Print("список: ")
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func (ol *OneList) SaveToFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := ol.head
	for curr != nil {
		_, err := file.WriteString(curr.Data + " ")
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (ol *OneList) LoadFromFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	content, err := io.ReadAll(file)
	if err != nil {
		return err
	}

	values := strings.Fields(string(content))
	for _, val := range values {
		ol.AddToTail(val)
	}
	return nil
}

func (ol *OneList) SaveToBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := ol.head
	for curr != nil {
		strSize := uint64(len(curr.Data))
		err := binary.Write(file, binary.LittleEndian, strSize)
		if err != nil {
			return err
		}
		_, err = file.WriteString(curr.Data)
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (ol *OneList) LoadFromBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	for {
		var strSize uint64
		err := binary.Read(file, binary.LittleEndian, &strSize)
		if err != nil {
			if err == io.EOF {
				break
			}
			return err
		}

		buffer := make([]byte, strSize)
		_, err = file.Read(buffer)
		if err != nil {
			return err
		}
		ol.AddToTail(string(buffer))
	}
	return nil
}

File: lab3/zad1/cpp/onelist.cpp
#include <iostream>
#include <fstream>
#include "node.h"
#include "onelist.h"

using namespace std;

OneList::OneList() {
    head = nullptr;
    tail = nullptr;
}

OneList::~OneList() {
    OneListNode* curr = head;
    while (curr != nullptr) {
        OneListNode* next = curr->next;
        delete curr;
        curr = next;
    }
}

void OneList::addToHead(const string& val) {
    auto* newNode = new OneListNode(val);
    newNode->next = head;   
    head = newNode;
    if (tail == nullptr) {
        tail = newNode;
    }
}

void OneList::addToTail(const string& val) {
    auto *newNode = new OneListNode(val);
    newNode->next = nullptr;    
    if (tail == nullptr) {
        head = newNode;
        tail = newNode;
    } else {
        tail->next = newNode;
        tail = newNode; 
    }
}

void OneList::addAfterValue(const string& targetVal, const string& newVal) {
    OneListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }

    auto *newNode = new OneListNode(newVal);
    newNode->next = targetNode->next;
    targetNode->next = newNode;
    
    if (targetNode == tail) {
        tail = newNode;
    }
}

void OneList::addBeforeValue(const string& targetVal, const string& newVal) {
    OneListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }
    
    if (targetNode == head) {
        addToHead(newVal);
        return;
    }
    
    OneListNode* prevNode = head;
    while (prevNode != nullptr && prevNode->next != targetNode) {
        prevNode = prevNode->next;
    }

    if (prevNode == nullptr) {
        throw runtime_error("error");
    }

    auto *newNode = new OneListNode(newVal);
    newNode->next = targetNode;
    prevNode->next = newNode;
}

void OneList::removeFromHead() {
    if (head == nullptr) {
        throw runtime_error("error");
    }  
    OneListNode* temp = head;
    head = head->next;   
    if (head == nullptr) {
        tail = nullptr;
    }    
    delete temp;
}

void OneList::removeFromTail() {
    if (head == nullptr) {
        throw runtime_error("error");
    }  
    if (head == tail) {
        delete head;
        head = nullptr;
        tail = nullptr;
    } else {
        OneListNode* curr = head;
        while (curr->next != tail) {
            curr = curr->next;
        }     
        delete tail;
        tail = curr; 
        tail->next = nullptr;
    }
}

void OneList::removeAfterValue(const string& targetVal) {
    OneListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr || targetNode->next == nullptr) {
        throw runtime_error("error");
    }
    
    OneListNode* temp = targetNode->next;
    targetNode->next = temp->next;
    
    if (temp == tail) {
        tail = targetNode;
    }
    
    delete temp;
}

void OneList::removeBeforeValue(const string& targetVal) {
    OneListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr || targetNode == head) {
        throw runtime_error("error");
    }
    
    if (head->next == targetNode) {
        removeFromHead();
        return;
    }
    OneListNode* prevPrev = head;
    while (prevPrev->next != nullptr && prevPrev->next->next != targetNode) {
        prevPrev = prevPrev->next;
    }
    
    OneListNode* temp = prevPrev->next;
    prevPrev->next = targetNode;
    
    delete temp;
}

void OneList::removeByValue(const string& targetVal) {
    OneListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }
    
    if (targetNode == head) {
        removeFromHead();
        return;
    }

    OneListNode* prevNode = head;
    while (prevNode != nullptr && prevNode->next != targetNode) {
        prevNode = prevNode->next;
    }

    if (prevNode == nullptr) {
        throw runtime_error("error");
    }

    prevNode->next = targetNode->next;
    
    if (targetNode == tail) {
        tail = prevNode;
    }
    
    delete targetNode;
}

auto OneList::getElementByValue(const string &targetVal) const -> string
{
    OneListNode* targetNode = findByVal(targetVal);
    if (targetNode == nullptr) {
        throw runtime_error("error");
    }
    return targetNode->data;
}

auto OneList::findByVal(const string &val) const -> OneListNode *
{
    OneListNode* curr = head;   
    while (curr != nullptr) {
        if (curr->data == val) {
            return curr;  
        }
        curr = curr->next;
    }    
    return nullptr;
}

void OneList::print() const {
    if (head == nullptr) {
        cout << "error" << '\n';
        return;
    }
    
    OneListNode* curr = head;
    cout << "список: ";
    while (curr != nullptr) {
        cout << curr->data;
        if (curr->next != nullptr) {
            cout << " -> ";
        }
        curr = curr->next;
    }
    cout << '\n';
}

void OneList::saveToFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    OneListNode* curr = head;
    while (curr != nullptr) {
        file << curr->data << " ";
        curr = curr->next;
    }
}

void OneList::loadFromFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    string value;
    while (file >> value) {
        addToTail(value);
    }
}

void OneList::saveToBinaryFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    OneListNode* curr = head;
    while (curr != nullptr) {
        size_t strSize = curr->data.size();
        auto streamSize = static_cast<streamsize>(strSize);
        file.write(reinterpret_cast<const char*>(&strSize), sizeof(strSize));
        file.write(curr->data.c_str(), streamSize);
        curr = curr->next;
    }
}

void OneList::loadFromBinaryFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    while (file.peek() != EOF) {
        size_t strSize;
        file.read(reinterpret_cast<char*>(&strSize), sizeof(strSize));
        if (file.eof()){ 
            break;
        }
        auto streamSize = static_cast<streamsize>(strSize);
        char* buffer = new char[strSize + 1];
        file.read(buffer, streamSize);
        buffer[strSize] = '\0';
        addToTail(string(buffer));
        delete[] buffer;
    }
}